### PR TITLE
Trial fix (updated) for customer TSD errors from March 1, 2017

### PR DESCRIPTION
**Pull request by pepper-ankan**
A first trial.  Please see [this](https://docs.google.com/document/d/1w27Us6UV7lYh8i6ExKEOgyrR-reO-i3ZskLQnqOSXSA) for details.
**NOTE**:
- I don't intend to merge this in yet, not until we have tried it out.
- I'll try it out on test.main (yes, on PD-hosted dashboards, but at least make sure this doesn't completely break that).
- The value of `tsd.core.preload_uid_cache_before_serving` will be set to "false" if we do plan to use this: it's set to "true" for ease of testing right now.

### CODE REVIEW

**Comment by ssuchter** [comment]
> The value of tsd.core.preload_uid_cache_before_serving will be set to "false" if we do plan to use this: it's set to "true" for ease of testing right now.

It seems like to me that if this works, we'd want the default to be that we *do* wait to serve before preloading the uid cache, right?

**Comment by pepper-ankan** [comment]
Yes, but I thought we'd set that *explicitly* in the config file, to remind
ourselves ... what do you think?

Cheerio!
-Ankan-

**Comment by ssuchter** [comment]
You're right - since it's ansible-ized, making it explicit in the configs that are laid out is a fine thing.

File: src/uid/UniqueId.java [hunk 1]
```diff
@@ -798,12 +816,20 @@ public Object call(final ArrayList<ArrayList<KeyValue>> rows) {
...
+          maybeLog("Done: got enough suggestions (" + suggestions.size()
+              + "); returning suggestions ...");
           return suggestions;
         }
         row.clear();  // free()
       }
       return search();  // Get more suggestions.
     }
+
+    private void maybeLog(String message) {
```
Comment (ssuchter) [inline src/uid/UniqueId.java]: I don't love how maybeLog doesn't indicate any of *why* it's maybe. Could it be named something like logIfPreloading? Or maybe if the only point of isUidCachePreload is to increase logging level, we could just name it 'extraLogging' and then keep this as maybeLog?

**Comment by pepper-ankan** [comment]
Yup, I also thought so (your first point) after pushing this to github ...
will change it (either first or second way) ...

Cheerio!
-Ankan-

**Comment by pepper-ankan** [comment]
Checking this in soon.  I'm leaving the value of `tsd.core.preload_uid_cache_before_serving` set to "true" for now; if we adopt this fix permanently, the default can be "false", and we can control it explicitly in the config file.

**Pull request merged**